### ISSUE TITLE
[SIP-96] Proposal for adding a notification feature

### ISSUE THREAD

**Issue by mattitoo** [apache/superset#24271]
**Motivation**
As maintainer of a Superset Instance, I want to be able to message my users, so I can inform them about possible downtimes, data troubles or other causes.
This can of course be achieved in various channels (Mail, Slack, Teams, etc) but it also makes sense to do this in the actual tool used, in this case right inside Superset.

**Proposed Change**
Luckily, Superset already has a messaging capability built in: so-called Toasts, which are displayed after certain actions like saving, refreshing dashboards and so on.
We have created a “management interface” for such toasts, in which new messages (we call them notifications) can be configured:

<img width="1561" alt="Bildschirmfoto 2023-05-30 um 15 37 25" src="https://github.com/apache/superset/assets/104143980/973c9869-0076-4feb-88b6-ffac2f63bb42">

<img width="833" alt="Bildschirmfoto 2023-05-30 um 15 37 39" src="https://github.com/apache/superset/assets/104143980/fd141bed-8b3a-4148-bd73-4fcf501bdd76">

Following parameters can be set:
·       Active
·       Name or Summary
·       Category (currently 3 categories, these manage the color of the notification: Info=blue, Warning=yellow, Alert=red)
·       Message: The message to be displayed.
·       Time Range: The range in which the notification should be displayed
·       Use daily timeframe: If active, the notification is only shown when the current time is inside the start and end time
·       Re-trigger: Minimum of 5 Minutes to reduce the load on the database due to querying for new messages for each user
·       Duration: How long until the notification disappears (it can always be closed via the ‘x’)
·       Roles: Which roles should see that notification

For each user, the internal database is queried at login and at a certain interval to see if she/he has new notifications pending.

Example notification in action:
<img width="311" alt="Bildschirmfoto 2023-05-30 um 15 57 20" src="https://github.com/apache/superset/assets/104143980/3583f943-21f3-4b36-8947-f1adf3c1fb0d">

NOTE: This should be behind a feature flag like “ENABLE_UI_NOTIFICATIONS”

**New or Changed Public Interfaces**
Extends internal database to include two additional schemas: public.notification and public.notification.roles

<img width="392" alt="Bildschirmfoto 2023-05-30 um 15 49 42" src="https://github.com/apache/superset/assets/104143980/bf7f52a5-e4d6-4e4d-8f4c-b3b3b0b00773">

Extends API to include endpoints for management of Notifications

<img width="1151" alt="Bildschirmfoto 2023-05-30 um 15 51 10" src="https://github.com/apache/superset/assets/104143980/47d24754-aa9d-43df-bc54-1626f664c1a5">

Adds an entry in the Settings menu

<img width="316" alt="Bildschirmfoto 2023-05-30 um 15 56 22" src="https://github.com/apache/superset/assets/104143980/c4a80d8d-b803-49e6-98a9-5aa846395d41">

Adds new permission for notification management

<img width="1150" alt="Bildschirmfoto 2023-05-30 um 15 57 09" src="https://github.com/apache/superset/assets/104143980/1365f7c1-6d3e-4920-8bd9-1d300b1ac707">

**Migration Plan and Compatibility**
Database migration necessary to include new tables

**Rejected Alternatives**
Using another method of displaying notifications, e.g. as a banner on top of Superset or below the menu bar. Rejected in favor of using existing and proven methods already built in.

**Comment by pankajsoni22** [comment]
I agree with the thought process, it's definitely a good step make this tool better and more useful.

**Comment by rusackas** [comment]
Thanks for the SIP! Numbering this as SIP-96. Please make sure to submit it on a [DISCUSS] thread on the Superset @dev mailing list, so we can start progressing it toward a vote. Thanks!

**Comment by michael-s-molina** [comment]
Thanks for the SIP! Some thoughts:

- It would be a good idea to get @kasiazjc's input on the screens.
- I really like @villebro's idea of different types of notifications. A banner would be helpful for critical messages that need to persist while an event is happening.
- Regarding the HTML sanitization, you can always extend the default schema with `HTML_SANITIZATION_OVERRIDES` and keep it safe with a well defined Content Security Policy (CSP). The important part here is that notifications should respect these settings.

**Comment by villebro** [comment]
I agree with @michael-s-molina 's proposal to use `HTML_SANITIZATION` and `HTML_SANITIZATION_SCHEMA_EXTENSIONS` (I believe that's what they're called?). I feel this needs to be part of the initial PR, as it's already an established pattern in the codebase. Regarding toasts vs banners, I'm fine with @mattitoo 's proposal to keep the scope limited to toasts for now. Let's just make sure the initial body of work can stand on its own (=so the project isn't stuck with more dead wood if development interest simmers) and we don't paint ourselves into a corner in a way which makes it difficult to extend notifications beyond toasts. And to distribute the work, I'm quite happy to pick up the task of creating banner notifications after the initial feature is ready if nobody else is volunteering for it.

One more thing that came to mind - should we support Jinja templating in these notifications? This will make it possible to make dynamic messages, e.g. "You only have 2 days left to migrate X to Y" instead of "Please migrate X to Y by Z". This may also be something worth considering for the other fields, like Category: It's an `Info` when there's more than 7 days left, becomes and `Alert` at 3 days and `Danger` when there's less than 2 days. Also, I brought this up in my initial message, but I think we should also consider adding support for snoozing and acknowledging messages.

**Comment by villebro** [comment]
I agree, let's not overload the initial work, but let's make sure we keep these in mind early on so the initial design can accommodate iterative improvements later. To summarize; let's leave as much room for extensibility as possible.

**Comment by michael-s-molina** [comment]
> I agree, let's not overload the initial work, but let's make sure we keep these in mind early on so the initial design can accommodate iterative improvements later. To summarize; let's leave as much room for extensibility as possible.

@mattitoo @villebro Exactly. The important part is that the foundation considers these future points of development. As we can see by the comments, the proposed feature seems to have great potential. Thanks again for the proposal @mattitoo and for driving its development. I'll gladly help reviewing the work and maybe contributing as well.

**Comment by rusackas** [comment]
Hi @mattitoo (and everyone) - I just wanted to check in on the status here. If we intend to move forward with this effort, someone will need to start a "DISCUSS" thread on the Apache Superset dev mailing list. After that, we can carry it forward to a vote. Ping me on Slack if help is needed.

**Comment by rusackas** [comment]
Hi @mattitoo - just wondering if there's any hope of moving this forward. I can kick off the [DISCUSS] thread for you on the mailing list if you'd like.

**Comment by MialLewis** [comment]
Just commenting to say that this feature would be much appreciated!

**Comment by vmbaraiya** [comment]
+1, Adding this feature would be highly beneficial for running Superset workloads in production environments.

**Comment by rusackas** [comment]
@mattitoo any ideas of timing here? Would love to see this move forward through the process (starting with a Discuss thread)

**Comment by sudhakar-datta** [comment]
It would be amazing if there are plans for making into a particular release and timeframe. This feature will be one of the most useful feature in a production workflow.

**Comment by rusackas** [comment]
This may be closed due to inactivity at some point, unless someone wants to pick up the torch and carry this through to a [DISCUSS] thread. Questions on how that works? Happy to help.

**Comment by villebro** [comment]
Sad to see this stall, this would be a really nice feature 🙁

**Comment by rusackas** [comment]
Spoke to @mattitoo on Slack, and there's still interest to rekindle this when time allows.

**Comment by rusackas** [comment]
By normal process, we should close this as inactive... is there still any interest in getting this through? I'm still a fan!

**Comment by sfirke** [comment]
@amaannawab923 has just made major PRs so I think it's reasonable to move this forward and re-assign to him. I found the original DISCUSS email that went out to the project listserve: https://lists.apache.org/thread/12ttbmmg9qogc4b1tx9wpj04l2y2gwxw One might say this has been open for discussion long enough, and gotten enough input on this thread, that we could move to a vote. Or we could bump it and give it a little time in case anything has changed. Either way.

**Comment by TechAuditBI** [comment]
Hey! Just saw this SIP. 
As a suggestion, maybe it would be easier and more clear to open a new SIP for further progress?

Speaking of a SIP itself, I do think that in order to make this feature more resilient and somewhat "up to date" it should be transformed into some form of a notification centre. Which would store all notifications received by user, mark read/unread ones and so on. Also packing notifications into a separate "module" would allow easier further development and upgrades. F.e. displaying changes in a new version for users in a form of a notification with some illustrations and maybe even gifs for easier understanding.

**Comment by rusackas** [comment]
I'm open to a new SIP or rekindling this one. It really depends on whether you want to take this exact approach or propose/follow a different path.

**Comment by rusackas** [comment]
I'm leaving this open until someone adopts the work, because I love it :)

**Comment by rusackas** [comment]
@mistercrunch this might (slightly) relate to the flash message stuff you were removing from FAB. Just since these are server-generated messages that reach the frontend. Maybe there's a pattern to reuse here somewhere :)

**Comment by mistercrunch** [comment]
Defintely want to kill `flash` specifically as its built to serve a server-side app (not designed for SPA), but the frontend portion could be retrofitted, maybe. I'm thinking we need proper/modern notif bell + drawers, and a more formal framework / service for it on the backend. Would make this a new effort as opposed to retrofitting what we have.

**Comment by mistercrunch** [comment]
In-scope/related is a sticky announcement banner that's a special kind of notif that may be more in-your-face. "Superset will be down from 6AM to 9AM this Sunday 2025-10-31"

**Comment by rusackas** [comment]
This hit my stale issue radar, but I'm not gonna close this one... I'll probably open the PR myself if nobody else volunteers. Just trying to get trough a bunch of other efforts first :)

**Comment by rusackas** [comment]
I'm itching to build this now, but the more I think of it, the more I realize this should be an Extension. Labeling it as such, and will look for a time/opportunity to create the extension interface. This might tie into turning Alerts/Reports into Extensions as well (output Extensions for Slack/Email/Webhooks/Discord/Teams/Whatever, and scheduler Extensions for Cron/SQL/Webhook/etc)
